carbon/__init__.py

```diff
@@ -61,39 +78,43 @@ def run(
             else 0.0,
             per_gb_power_watts=config.memory[dummy.mem_type]["per_gb_power_watts"],
         )
+        job_list = [dummy_job] * len(job_ids)
+        node_list = [dummy_node] * len(job_ids)
     else:
-        # Remove suffix to make IDs more uniform
-        id = job_id.split(".")[0]
-
-        if id.endswith("[]"):
-            raise UnsupportedJobType("array")
-
-        # Fetch job data from the cluster's job scheduler
-        job = Job.fromPBS(id)
-        node = Node.fromPBS(
-            job.node,
+        job_list = Job.from_PBS_bulk(job_ids, ignore_failed)
+        node_list = Node.from_PBS_bulk(
+            [job.node for job in job_list],
             {
                 "cpus": config.cpus,
                 "gpus": config.gpus,
                 "memory": config.memory,
             },
         )
-    # Calculate energy consumption
-    energy_consumed = job.calculate_energy(node, config.pue)
 
-    # Fetch carbon intensity at job start time or use a default value
-    if default_intensity:
-        intensity = 137.0  # gCO2/kWh, UK average over 2023 and 2024
-    else:
-        carbon_intensity = CarbonIntensity(job.starttime, region_id=config.region_id)
-        intensity = carbon_intensity.fetch()
-
-    # Calculate emissions
-    emissions = intensity * energy_consumed
-    return RunResult(
-        node=node,
-        emissions=emissions,
-        energy_consumed=energy_consumed,
-        job=job,
-        carbon_intensity=intensity,
-    )
+    result_list = []
+
+    for job, node in zip(job_list, node_list):
+        # Calculate energy consumption
+        energy_consumed = job.calculate_energy(node, config.pue)
+
+        # Fetch carbon intensity at job start time or use a default value
+        if default_intensity:
+            intensity = 137.0  # gCO2/kWh, UK average over 2023 and 2024
+        else:
+            carbon_intensity = CarbonIntensity(job.starttime, config.region_id)
+            intensity = carbon_intensity.fetch()
+
+        # Calculate emissions
+        emissions = intensity * energy_consumed
+
+        result_list.append(
+            RunResult(
+                node=node,
+                emissions=emissions,
+                energy_consumed=energy_consumed,
+                job=job,
+                carbon_intensity=intensity,
+            )
+        )
+
+    return result_list, len(job_ids) - len(result_list)
```

Comment:
Suggest avoiding returning tuples like this with different datastructures. Would recommend adding a dedicated dataclass with named fields.